### PR TITLE
Allow Tensor-likes in torch.autograd.gradcheck

### LINKED ISSUE #42942: torch.autograd.gradcheck support for Tensor-like types (__torch_function__).
## 🚀 Feature
[gradcheck.py#L264](https://github.com/pytorch/pytorch/blob/master/torch/autograd/gradcheck.py#L264) explicitly requires its inputs to be of type torch.Tensor and thus doesn't work for Tensor-like objects.

## Motivation

Tensor-like objects might implement operations with autograd support. gradcheck helps check their correctness. Extending torch.autograd.gradcheck with support for Tensor-like types makes it easier for users to test their code.

## Pitch

This feature requests asks to extend gradcheck to also accept Tensor-like types and proposes using the presence of ```__torch_function__``` as a determinant, as we [usually do](https://github.com/pytorch/pytorch/blob/3d46e02ea1f8951bcef167c33bad0217ec9da980/torch/csrc/utils/python_arg_parser.cpp#L182). The rest of the code should work fine then, assuming the user has reasonable implementations for the used features. Potentially we could guard using methods such as "is_sparse" based on whether the given type is Tensor-like and actually has that method defined. This might ease the implementation of Tensor-like objects, because it doesn't require the user to implement those. We could signal this with a warning as well as we do when the user doesn't use float64 as an input dtype, etc.

## Alternatives

An alternative as a user is to copy-paste and modify this code for their own purposes.

## Additional context

I'm using this in context of the NestedTensor project. I labeled this as "module: numpy" based on the labels used in https://github.com/pytorch/pytorch/issues/24015.

cc @ezyang @SsnL @albanD @zou3519 @gqchen @mruberry @rgommers @VitalyFedyunin @hameerabbasi

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #42942

Re-do of #43877.

### CODE REVIEW

File: torch/autograd/gradcheck.py [hunk 1]
```diff
@@ -277,7 +277,8 @@ def gradcheck(
             identical inputs through the differentiation, the results must either match
             exactly (default, 0.0) or be within this tolerance.
         check_undefined_grad (bool, options): if True, check if undefined output grads
-            are supported and treated as zeros
+            are supported and treated as zeros. Defaults to ``True`` for ``Tensor`` outputs
```
Comment (albanD) [inline torch/autograd/gradcheck.py:280]: nit: should this say inputs? Or should the check below check the outputs? I think the two should be consistent.

**Review by albanD — COMMENTED** [review]
Looks good.
Note that another option can be to leave the current default and just say:
```
if True, check if undefined output grads are supported and treated as zeros. This flag is ignored for non-``Tensor`` outputs.
```

**Comment by hameerabbasi** [comment]
> Note that another option can be to leave the current default and just say:

I'd prefer not to ignore a flag when it is set:
```
>>> import this
...
Explicit is better than implicit.
...
```

File: torch/autograd/gradcheck.py (hunk 1 above)
Comment (hameerabbasi) [reply torch/autograd/gradcheck.py]: Right, done!

**Review by albanD — COMMENTED** [review]
I don't think the doc you mention is accurate. You actually default to True if *all* inputs are Tensors and False otherwise.
I think there is an issue with that:
- This will disable this for other functions that pass arguments like `(t1, t2, float_eps)` and I'm pretty sure we have some of these.
- We don't have to disable this if any is non-Tensor. We just don't want to use this for non-Tensor objects.

Do you think that modifying the line below to:
```python
outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs)) if isinstance(o, torch.Tensor)]]
```
Would work?

And you can change the doc accordingly to mention that the flag only affect Tensor objects.

**Comment by hameerabbasi** [comment]
> Do you think that modifying the line below to:
> 
> ```python
> outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs)) if isinstance(o, torch.Tensor)]]
> ```
> 
> Would work?
> 
> And you can change the doc accordingly to mention that the flag only affect Tensor objects.

That's a fair point; done!

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -462,7 +462,10 @@ def check_undefined_grad_support(output_to_check):
                 return True
 
             # All backward functions must work properly if all output grads are undefined
-            outputs_to_check = [[torch._C._functions.UndefinedGrad()(o) for o in _differentiable_outputs(func(*tupled_inputs))]]
+            outputs_to_check = [[
```
Comment (albanD) [inline torch/autograd/gradcheck.py:465]: nit: Can you add an extra comment here:
"The isinstance check below allows to filter out Tensor-like objects that are returned by `_differentiable_outputs`."

**Review by albanD — COMMENTED** [review]
LGTM
Just one small comment.

Out of curiosity. Does this work with `gradgradcheck` as well?

File: torch/autograd/gradcheck.py (hunk 2 above)
Comment (hameerabbasi) [reply torch/autograd/gradcheck.py:465]: Added below.

**Comment by hameerabbasi** [comment]
> Out of curiosity. Does this work with `gradgradcheck` as well?

It does, added a test.

**Review by albanD — APPROVED** [review]
Perfect!
Thanks for the updates.

**Comment by hameerabbasi** [comment]
The remaining test failure appears to be a persistent network issue, and unrelated to this PR.

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**